Logging 2.0

TYWong99 opened the pull request:
This Pull Request includes implementation of the Logging feature, aiming to replace `print()` for console outputs. The affected files also include miscellaneous changes for better organisation, such as dividers between "sections". Below are the affected files:

- `log_utils.py`
- `cli_collide.py`
- `pattern_matcher.py`
- `level_playdo.py` (technically unchanged, since all `print()` were disabled)

`cli_test.py` is also included, and can be used as file template if needed.

TYWong99 committed:
Logging 2.0

This Pull Request includes implementation of the Logging feature, aiming to replace `print()` for console outputs. The affected files also include miscellaneous changes for better organisation, such as dividers between "sections". Below are the affected files:

- `log_utils.py`
- `cli_collide.py`
- `pattern_matcher.py`
- `level_playdo.py`
(technically unchanged, since all `print()` were disabled)

`cli_test.py` is also included, and can be used as file template if needed.

qtran122 commented on logic/common/log_utils.py:
```diff
@@ -0,0 +1,111 @@
+'''Public Functions'''
+
+def Must(s):
+    '''Logs messages no matter the verbosity leve'''
```
*level

qtran122 commented on logic/common/log_utils.py:
```diff
@@ -0,0 +1,111 @@
+    log.SetVerbosityLevel(1)
+    log.Info("Hello World!")
+
+# Verbosity level is usually first set in CLI file once during initialisation
```
Feels redundant to have a '#' inside of a multi-line comment. let's get rid of it

qtran122 commented on logic/common/log_utils.py:
```diff
@@ -0,0 +1,111 @@
+''' Utility functions for printing and logging
```
''' <newline here>
Utility functions for printing and logging

qtran122 commented on logic/common/log_utils.py:
```diff
@@ -0,0 +1,111 @@
+#--------------------------------------------------#
+'''Public Functions'''
+
+def Must(s):
```
rename the variable 's' to 'msg'

qtran122 commented on logic/pattern/pattern_matcher.py:
```diff
@@ -1,20 +1,44 @@
-'''A convenience module that...
+''' PatternMatcher is a convenience module that scans the entire level for matching patterns.
```
'''
PatternMatcher is a convenience module that scans the entire level for matching patterns.

qtran122 commented on logic/pattern/pattern_matcher.py:
```diff
@@ -1,20 +1,44 @@
+    pattern_matcher_bb = PM.PatternMatcher()
+    for pattern in _LIST_PATTERN_BB:
+        pattern_matcher_bb.LoadPattern(pattern_root + pattern + ".xml")
+    pattern_matcher_bb.FindAndCreate(playdo, "_BB", "collisions_BB", allow_overlap = False)
```
`USAGE EXAMPLE:`
`import logic.pattern.pattern_matcher as PM`
`    pattern_matcher_bb = PM.PatternMatcher()`
`    pattern_matcher_bb.LoadPattern("breakable_blocks_pattern.xml")`
`    pattern_matcher_bb.FindAndCreate(playdo, "_BB_input", "collisions_BB_output", allow_overlap = False)`

I'd rewrite the above comments to be like the above - first for consistency in indentation. And second, to be more compact. Remember, the usage is meant to be representative and not a comprehensive example.

qtran122 commented on logic/common/log_utils.py:
```diff
@@ -0,0 +1,111 @@
+
+
+
+def _WriteHeaderToLog(log_lv_num):
+    print_msg = ""
+    print_msg += "\n" + DEFAULT_DIVIDER + "\n"    # Divider before meta data
+    print_msg += "  "                             # Indentation
+    print_msg += os.getcwd() + "\n"               #  Current OS
+    print_msg += "  "                             # Indentation
+    print_msg += "Verbosity = " + str(log_lv_num) #  Verbosity Level
+    print_msg += "" # TODO DateTime
+    print_msg += ""
+    print_msg += "\n" + DEFAULT_DIVIDER           # Divider between meta data & tool content
+    print_msg += "\n"
+
+    # Header is logged regardless of verbosity
+    logging.warning(print_msg)
```
Let's go for a reasonable amount of readability and compactness. Suggest:

`def _WriteHeaderToLog(log_lv_num):`
`    print_msg  = "\n" + DEFAULT_DIVIDER                 # Divider before meta data`
`    print_msg += "\n  " + os.getcwd()                   # Current OS`
`    print_msg += "\n  Verbosity = " + str(log_lv_num)   # Verbosity Level`
`    print_msg += "\n" + DEFAULT_DIVIDER + "\n"          # Divider between meta data & tool content`
`    #print_msg += "" # TODO DateTime`
`    logging.warning(print_msg)  # Header is logged regardless of verbosity`

qtran122 commented on cli_collide.py:
```diff
@@ -1,10 +1,14 @@
 ''' Command-Line Tool to auto-add ground, BBs, crystals, and skell reefs to a level. 
```
`''' `
`Command-Line Tool to auto-add ground, BBs, crystals, and skell reefs to a level. `

Let's put the first line down, so that it's not "indented"

TYWong99 committed:
Logging 2.1

Applied the requested changes. Additionally,

- Header Descriptions re-ordered, now <USAGE> section is always after text descrpition
- `cli_collide.py` Descrpitions changed
